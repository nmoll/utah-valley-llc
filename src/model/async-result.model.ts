export interface Success<T> {
  type: "success";
  data: T;
}

export interface Error {
  type: "error";
}

export type AsyncResult<T> = Success<T> | Error;
