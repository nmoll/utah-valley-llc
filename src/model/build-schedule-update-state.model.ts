export const ScheduleUpdateActionType = {
  SwapHosts: "Swap hosts",
  "Swap pianists": "Swap pianists",
  "Swap bible class": "Swap bible class",
  "Add new service": "Add new service",
  "Cancel service": "Cancel service",
  "Add me to rotation": "Add me to rotation",
  "Remove me temporarily": "Remove me temporarily",
  "Remove me permanently": "Remove me permanently",
} as const;

export const scheduleUpdateActions = Object.values(ScheduleUpdateActionType);

export type ScheduleUpdateAction =
  typeof ScheduleUpdateActionType[keyof typeof ScheduleUpdateActionType];

interface ChooseAction {
  type: "Choose Action";
}

export type BuildScheduleUpdateState = ChooseAction;
