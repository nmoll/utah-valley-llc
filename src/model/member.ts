import { Dayjs } from "dayjs";

export interface Member {
  name: string;
  active: Dayjs;
  inactive?: {
    start: Dayjs;
    end: Dayjs | null;
  }[];
}
