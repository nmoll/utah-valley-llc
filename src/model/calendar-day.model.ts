import { CalendarEvent } from "./calendar-event.model";

export interface CalendarDay {
  date: Date;
  isCurrentMonth: boolean;
  isToday: boolean;
  event: CalendarEvent | null;
}
