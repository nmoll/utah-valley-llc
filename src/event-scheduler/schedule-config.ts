import { Member } from "../model/member";
import { ScheduleUpdate } from "../model/schedule-update.model";

export interface ScheduleConfig {
  hosts: Member[];
  pianists: Member[];
  bibleClassLeaders: Member[];
  serviceDirectors: Member[];
  scheduleUpdates: ScheduleUpdate[];
}
