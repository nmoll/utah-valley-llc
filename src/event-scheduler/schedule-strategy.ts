import { Dayjs } from "dayjs";
import { Member } from "../model/member";
import { ScheduleUpdate } from "../model/schedule-update.model";

export class ScheduleEventStrategyFactory {
  constructor(private scheduleUpdates: ScheduleUpdate[]) {}

  getAll(): ScheduleEventStrategy[] {
    return [
      new UpdatedEventStrategy(this.scheduleUpdates),
      new SundayMorningServiceStrategy(),
      new SongServicesStrategy(),
      new BibleClassStrategy(),
    ];
  }

  getStrategy(
    date: Dayjs,
    options: { exclude?: ScheduleEventStrategy } = {}
  ): ScheduleEventStrategy | null {
    const strategy = this.getAll().find((strategy) => {
      if (
        options.exclude &&
        options.exclude.constructor === strategy.constructor
      ) {
        return false;
      }
      return strategy.matches(date);
    });
    return strategy ?? null;
  }
}

export interface ScheduleStrategy {
  matches(date: Dayjs): boolean;
}

export interface ScheduleEventStrategy extends ScheduleStrategy {
  getDescription(date: Dayjs): string;
  getScheduleHostStrategy(date: Dayjs): ScheduleHostStrategy;
}

export class UpdatedEventStrategy implements ScheduleEventStrategy {
  constructor(private updates: ScheduleUpdate[]) {}

  matches(date: Dayjs): boolean {
    return !!this.getUpdate(date);
  }

  getDescription(date: Dayjs): string {
    const update = this.getUpdate(date);
    if (update?.changes?.description) {
      return update.changes.description;
    }
    const defaultStrategy = this.getDefaultStrategy(date);
    if (defaultStrategy) {
      return defaultStrategy.getDescription(date);
    }

    return "";
  }

  getScheduleHostStrategy(date: Dayjs): ScheduleHostStrategy {
    const defaultStrategy = this.getDefaultStrategy(date);
    if (defaultStrategy) {
      return defaultStrategy.getScheduleHostStrategy(date);
    }
    return new ScheduleEveningHostStrategy();
  }

  private getDefaultStrategy(date: Dayjs): ScheduleEventStrategy | null {
    return new ScheduleEventStrategyFactory(this.updates).getStrategy(date, {
      exclude: this,
    });
  }

  private getUpdate(date: Dayjs): ScheduleUpdate | undefined {
    return this.updates.find((update) => update.date.isSame(date, "day"));
  }
}

export class BibleClassStrategy implements ScheduleEventStrategy {
  matches(date: Dayjs): boolean {
    return date.day() === 3 && date.date() > 7;
  }
  getDescription() {
    return "7pm Bible Class";
  }
  getScheduleHostStrategy(): ScheduleHostStrategy {
    return new ScheduleEveningHostStrategy();
  }
}

export class SongServicesStrategy implements ScheduleEventStrategy {
  matches(date: Dayjs): boolean {
    return date.day() === 3 && date.date() <= 7;
  }
  getDescription(): string {
    return "7pm Song Services";
  }
  getScheduleHostStrategy(): ScheduleHostStrategy {
    return new ScheduleEveningHostStrategy();
  }
}

export class SundayMorningServiceStrategy implements ScheduleEventStrategy {
  matches(date: Dayjs): boolean {
    return date.day() === 0;
  }
  getScheduleHostStrategy(): ScheduleHostStrategy {
    return new ScheduleMorningHostStrategy();
  }
  getDescription(): string {
    return "10:30 Service";
  }
}

export interface ScheduleHostStrategy extends ScheduleStrategy {
  offsetInitialHost(): boolean;
  getRotationOffset(hosts: Member[]): Member | null;
}

export class ScheduleMorningHostStrategy implements ScheduleHostStrategy {
  matches(date: Dayjs): boolean {
    return date.day() === 0;
  }
  offsetInitialHost(): boolean {
    return false;
  }
  getRotationOffset(): Member | null {
    return null;
  }
}

export class ScheduleEveningHostStrategy implements ScheduleHostStrategy {
  matches(date: Dayjs): boolean {
    return date.day() !== 0;
  }
  offsetInitialHost(): boolean {
    return true;
  }

  /**
   * Offsets the rotation to the middle host so morning
   * and evening rotations are different
   */
  getRotationOffset(hosts: Member[]): Member | null {
    return hosts[Math.floor((hosts.length - 1) / 2)];
  }
}

export class ScheduleMultiplePianistsStrategy implements ScheduleStrategy {
  constructor(private updates: ScheduleUpdate[]) {}

  matches(date: Dayjs): boolean {
    return this.updates.some(
      (update) => update.date.isSame(date, "day") && update.twoPianists
    );
  }
}
