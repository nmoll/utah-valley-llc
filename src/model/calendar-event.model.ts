import { Dayjs } from "dayjs";
import { Member } from "./member";

export interface CalendarEvent {
  date: Dayjs;
  description: string;
  host: Member;
  bibleClassLeader: Member | null;
  pianists: Member[];
}

export type CalendarEvents = Record<string, CalendarEvent>;
