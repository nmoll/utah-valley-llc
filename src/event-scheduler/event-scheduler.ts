import { Dayjs } from "dayjs";
import { CalendarEvent } from "../model/calendar-event.model";
import { Member } from "../model/member";
import { ScheduleUtil } from "../util/schedule.util";
import { ScheduleConfig } from "./schedule-config";
import {
  BibleClassStrategy,
  ScheduleEventStrategyFactory,
  ScheduleMultiplePianistsStrategy,
} from "./schedule-strategy";

export class EventScheduler {
  constructor(private config: ScheduleConfig) {}

  private scheduleEventStrategyFactory = new ScheduleEventStrategyFactory(
    this.config.scheduleUpdates
  );

  schedule(date: Dayjs): CalendarEvent | null {
    const update = this.config.scheduleUpdates.find((update) =>
      update.date.isSame(date, "day")
    );

    if (update?.cancelled) {
      return null;
    }

    const scheduleStrategy =
      this.scheduleEventStrategyFactory.getStrategy(date);
    if (!scheduleStrategy) {
      return null;
    }

    const description = scheduleStrategy.getDescription(date);
    const host = this.scheduleHost(date);
    const pianists = this.schedulePianists(date);
    const bibleClassLeader = this.scheduleBibleClassLeader(date);

    if (!host) {
      return null;
    }

    return {
      date,
      description,
      host,
      pianists,
      bibleClassLeader,

      ...update?.changes,
    };
  }

  scheduleAll(start: Dayjs, end: Dayjs): CalendarEvent[] {
    return ScheduleUtil.getDatesBetween(start, end)
      .map((date) => this.schedule(date))
      .filter((event): event is CalendarEvent => !!event);
  }

  private scheduleHost(date: Dayjs): Member | null {
    const eventDates = this.getEventDates(date);

    return ScheduleUtil.scheduleMember(this.config.hosts, eventDates);
  }

  private schedulePianists(date: Dayjs): Member[] {
    const schedulePianistStrategy = new ScheduleMultiplePianistsStrategy(
      this.config.scheduleUpdates
    );
    const eventDates = this.getEventDates(date);
    const pastEventsWithMultiplePianists = eventDates
      .filter((d) => d.isBefore(date))
      .filter(ScheduleUtil.dateFilter([schedulePianistStrategy]));

    const allEvents = eventDates
      .concat(pastEventsWithMultiplePianists)
      .sort((a, b) => a.unix() - b.unix());

    const pianist = ScheduleUtil.scheduleMember(
      this.config.pianists,
      allEvents
    );

    if (!pianist) {
      return [];
    }

    const result = [pianist];

    if (schedulePianistStrategy.matches(date)) {
      result.push(
        ScheduleUtil.getNextActiveMember(pianist, this.config.pianists, date)
      );
    }
    return result;
  }

  private scheduleBibleClassLeader(date: Dayjs): Member | null {
    const bibleClassStrategy = new BibleClassStrategy();
    if (!bibleClassStrategy.matches(date)) {
      return null;
    }

    const eventDates = this.getEventDates(date).filter(
      ScheduleUtil.dateFilter([bibleClassStrategy])
    );

    return ScheduleUtil.scheduleMember(
      this.config.bibleClassLeaders,
      eventDates
    );
  }

  /**
   * Returns all dates which match any event scheduling strategy
   * between the scheduling start date and the given date.
   *
   * The scheduling start date is determined by the
   * first host's active date
   */
  private getEventDates(date: Dayjs): Dayjs[] {
    return ScheduleUtil.getDatesBetween(this.config.hosts[0].active, date)
      .filter((date) => !this.isCancelled(date))
      .filter(
        ScheduleUtil.dateFilter(this.scheduleEventStrategyFactory.getAll())
      );
  }

  private isCancelled(date: Dayjs): boolean {
    return this.config.scheduleUpdates.some(
      (update) => update.date.isSame(date, "day") && update.cancelled
    );
  }
}
