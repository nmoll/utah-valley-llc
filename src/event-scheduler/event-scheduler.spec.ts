import * as dayjs from "dayjs";
import { CalendarEvent } from "../model/calendar-event.model";
import { Member } from "../model/member";
import {
  bibleClassLeaders,
  hosts,
  hostsByName,
  pianists,
  pianistsByName,
} from "../test/mock-data";
import { EventScheduler } from "./event-scheduler";
import { ScheduleConfig } from "./schedule-config";

const isCalendarEvent = (obj: any): boolean =>
  !!obj.date && !!obj.description && !!obj.host;

const serializeEvent = (event: CalendarEvent) =>
  `
${event.date.format("ddd MMM D")}: ${event.description}
  Host: ${event.host.name}
  Pianists: ${event.pianists.map((p) => p.name).join(", ")}
  Bible Class: ${event.bibleClassLeader?.name ?? ""}
`.trimStart();

expect.addSnapshotSerializer({
  test: isCalendarEvent,
  serialize: serializeEvent,
});

describe("EventScheduler", () => {
  const config: ScheduleConfig = {
    hosts: hosts,
    bibleClassLeaders: bibleClassLeaders,
    pianists: pianists,
    scheduleUpdates: [],
  };

  describe("scheduleAll()", () => {
    it("should schedule events between dates", () => {
      const scheduler = new EventScheduler(config);

      expect(
        scheduler.scheduleAll(dayjs("2022-01-02"), dayjs("2022-01-30"))
      ).toMatchSnapshot();
    });

    it("should apply description update to event", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-09"),
            changes: {
              description: "10:30 Communion",
            },
          },
        ],
      });

      expect(scheduler.schedule(dayjs("2022-01-09"))).toEqual<CalendarEvent>({
        date: dayjs("2022-01-09"),
        description: "10:30 Communion",
        bibleClassLeader: null,
        host: hostsByName["B"],
        pianists: [pianistsByName["C"]],
      });
    });

    it("should use default description not updated", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-09"),
            changes: {
              host: hostsByName["C"],
            },
          },
        ],
      });

      expect(
        scheduler.schedule(dayjs("2022-01-09"))?.description
      ).toEqual<string>("10:30am Service");
    });

    it("should not affect host rotation if replaced for existing event", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-09"),
            changes: {
              description: "10:30 Communion",
              host: hostsByName["D"],
            },
          },
        ],
      });

      expect(scheduler.schedule(dayjs("2022-01-02"))?.host).toEqual<Member>(
        hostsByName["A"]
      );
      expect(scheduler.schedule(dayjs("2022-01-09"))?.host).toEqual<Member>(
        hostsByName["D"]
      );
      expect(scheduler.schedule(dayjs("2022-01-16"))?.host).toEqual<Member>(
        hostsByName["C"]
      );
    });

    it("should use wednesday rotation if new evening event is scheduled without a specified host", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-07"),
            changes: {
              description: "New Event",
            },
          },
        ],
      });

      // Host before event is not affected
      expect(scheduler.schedule(dayjs("2022-01-05"))?.host).toEqual<Member>(
        hostsByName["C"]
      );
      // Next host on rotation is scheduled for new event
      expect(scheduler.schedule(dayjs("2022-01-07"))?.host).toEqual<Member>(
        hostsByName["D"]
      );
      // Sunday rotation is not affected
      expect(scheduler.schedule(dayjs("2022-01-09"))?.host).toEqual<Member>(
        hostsByName["B"]
      );
      // Wednesday rotation is adjusted
      expect(scheduler.schedule(dayjs("2022-01-12"))?.host).toEqual<Member>(
        hostsByName["A"]
      );
    });

    it("should use pianist rotation if new event is scheduled", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-08"),
            changes: {
              description: "New Event",
            },
          },
        ],
      });

      // Pianist before event is not affected
      expect(scheduler.schedule(dayjs("2022-01-05"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["B"]]);
      // Next pianist on rotation is scheduled for new event
      expect(scheduler.schedule(dayjs("2022-01-08"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["C"]]);
      // Sunday rotation is not affected
      expect(scheduler.schedule(dayjs("2022-01-09"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["D"]]);
      // Wednesday rotation is adjusted
      expect(scheduler.schedule(dayjs("2022-01-12"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["A"]]);
    });

    it("should allow two pianists to be scheduled", () => {
      const scheduler = new EventScheduler({
        ...config,
        scheduleUpdates: [
          {
            date: dayjs("2022-01-05"),
            changes: {},
            twoPianists: true,
          },
          {
            date: dayjs("2022-01-10"),
            changes: {
              description: "Special Service 1",
              host: hostsByName["A"],
            },
            twoPianists: true,
          },
          {
            date: dayjs("2022-01-11"),
            changes: {
              description: "Special Service 2",
            },
          },
        ],
      });

      expect(scheduler.schedule(dayjs("2022-01-02"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["A"]]);

      expect(scheduler.schedule(dayjs("2022-01-05"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["B"], pianistsByName["C"]]);

      expect(scheduler.schedule(dayjs("2022-01-09"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["D"]]);

      expect(scheduler.schedule(dayjs("2022-01-10"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["A"], pianistsByName["B"]]);

      expect(scheduler.schedule(dayjs("2022-01-11"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["C"]]);

      expect(scheduler.schedule(dayjs("2022-01-12"))?.pianists).toEqual<
        Member[]
      >([pianistsByName["D"]]);
    });
  });
});
