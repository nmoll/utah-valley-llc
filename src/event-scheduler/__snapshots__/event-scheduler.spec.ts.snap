// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventScheduler scheduleAll() should cancel service and continue rotation after 1`] = `
Array [
  Sun Jan 2: 10:30am Service
  Host: Host A
  Pianists: Pianist A
  Bible Class: 
,
  Sun Jan 9: 10:30am Service
  Host: Host B
  Pianists: Pianist B
  Bible Class: 
,
  Wed Jan 12: 7pm Bible Class
  Host: Host C
  Pianists: Pianist C
  Bible Class: BCL A
,
  Sun Jan 16: 10:30am Service
  Host: Host C
  Pianists: Pianist D
  Bible Class: 
,
  Wed Jan 19: 7pm Bible Class
  Host: Host D
  Pianists: Pianist A
  Bible Class: BCL B
,
  Sun Jan 23: 10:30am Service
  Host: Host D
  Pianists: Pianist B
  Bible Class: 
,
  Wed Jan 26: 7pm Bible Class
  Host: Host A
  Pianists: Pianist C
  Bible Class: BCL C
,
  Sun Jan 30: 10:30am Service
  Host: Host A
  Pianists: Pianist D
  Bible Class: 
,
]
`;

exports[`EventScheduler scheduleAll() should schedule events between dates 1`] = `
Array [
  Sun Jan 2: 10:30am Service
  Host: Host A
  Pianists: Pianist A
  Bible Class: 
,
  Wed Jan 5: 7pm Song Services
  Host: Host C
  Pianists: Pianist B
  Bible Class: 
,
  Sun Jan 9: 10:30am Service
  Host: Host B
  Pianists: Pianist C
  Bible Class: 
,
  Wed Jan 12: 7pm Bible Class
  Host: Host D
  Pianists: Pianist D
  Bible Class: BCL A
,
  Sun Jan 16: 10:30am Service
  Host: Host C
  Pianists: Pianist A
  Bible Class: 
,
  Wed Jan 19: 7pm Bible Class
  Host: Host A
  Pianists: Pianist B
  Bible Class: BCL B
,
  Sun Jan 23: 10:30am Service
  Host: Host D
  Pianists: Pianist C
  Bible Class: 
,
  Wed Jan 26: 7pm Bible Class
  Host: Host B
  Pianists: Pianist D
  Bible Class: BCL C
,
  Sun Jan 30: 10:30am Service
  Host: Host A
  Pianists: Pianist A
  Bible Class: 
,
]
`;
